clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -6471,16 +6471,32 @@ void SYCLIntegrationHeader::emit(raw_ostream &O) {
       O << "extern \"C\" ";
     std::string ParmList;
     bool FirstParam = true;
+    Policy.SuppressDefaultTemplateArgs = false;
     for (ParmVarDecl *Param : K.SyclKernel->parameters()) {
       if (FirstParam)
         FirstParam = false;
       else
         ParmList += ", ";
-      ParmList += Param->getType().getCanonicalType().getAsString();
+      ParmList += Param->getType().getCanonicalType().getAsString(Policy);
     }
     FunctionTemplateDecl *FTD = K.SyclKernel->getPrimaryTemplate();
     Policy.SuppressDefinition = true;
     Policy.PolishForDeclaration = true;
+    Policy.FullyQualifiedName = true;
+    Policy.EnforceScopeForElaboratedTypes = true;
+
+    // SuppressDefaultTemplateArguments is a downstream addition that suppresses
+    // default template arguments in function declarations. It should be set to
+    // true to emit function declaration that won't cause any compilation errors
+    // when present in the integration header.
+    // SuppressDefaultTemplateArgs is a flag from community code which adds
+    // suppression of template arguments that match default template arguments
+    // in template-ids printing.
+    Policy.SuppressDefaultTemplateArguments = true;
+    // EnforceDefaultTemplateArgs is a downstream addition that forces printing
+    // template arguments that match default template arguments while printing
+    // template-ids, even if the source code doesn't referece them.
+    Policy.EnforceDefaultTemplateArgs = true;
```

Comment:
The names make this quite confusing. 
`Policy.SuppressDefaultTemplateArguments` and `Policy.EnforceDefaultTemplateArgs` sounds like they do opposite things but are both set to `true` here. I assume `EnforceDefaultTemplateArgs` is technically the opposite of community `SuppressDefaultTemplateArgs` ? What does `match default template arguments` mean?

Reply:
> I assume EnforceDefaultTemplateArgs is technically the opposite of community SuppressDefaultTemplateArgs ?

Technically, yes.

> What does match default template arguments mean?

A template argument in a template-id and with the same value as default argument in the template declaration that this template-id refers to.

Policy.SuppressDefaultTemplateArguments suppresses default template arguments from definition of template, i.e.
if function decl has
```
template <typename = int> void Imafreefunction(....) {...}
```
`typename = int` will be suppressed leaving simply `typename` in the integration header. Perhaps I should rename this flag to something like `Policy.SuppressDefaultTemplateArgumentsInDeclarations`? WDYT? It is a flag added by intel/llvm anyway.

Reply:
Thanks for explaining! I think you can leave it as-is